src/main/java/fi/aalto/cs/apluscourses/utils/FileDateFormatter.java

```diff
@@ -0,0 +1,86 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+import java.time.Clock;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+
+public class FileDateFormatter {
+  // the order of time units to check, ordered from most to least precise
+  private static final List<ChronoUnit> TimeUnits = Arrays.asList(
+          ChronoUnit.SECONDS,
+          ChronoUnit.MINUTES,
+          ChronoUnit.HOURS,
+          ChronoUnit.DAYS,
+          ChronoUnit.MONTHS,
+          ChronoUnit.YEARS
+  );
+
+  private static String formatWithTimeUnit(ZonedDateTime fileTime,
+                                           ZonedDateTime currentTime,
+                                           ChronoUnit timeUnit) {
+    long timeUnitCount = timeUnit.between(fileTime, currentTime);
+    String timeUnitText = timeUnit.toString().toLowerCase(Locale.ROOT);
+    if (timeUnitCount == 1) {
+      // remove the final pluralizing "s" from the time unit name
+      timeUnitText = timeUnitText.substring(0, timeUnitText.length() - 1);
+    }
+
+    return timeUnitCount + " " + timeUnitText + " ago";
+  }
+
+  private static String formatWithLargestTimeUnit(ZonedDateTime fileTime, Clock currentTimeClock) {
+    ZonedDateTime currentTime = ZonedDateTime.now(currentTimeClock);
+
+    if (ChronoUnit.SECONDS.between(fileTime, currentTime) <= 0) {
+      return "just now";
+    }
```

Comment:
Does this mean what I think that it means? :smile: As soon as one full second has passed "just now" isn't shown anymore?

Reply:
It could as well be "0 seconds ago" if that would make the logic simpler.

Reply:
Or the seconds option could perhaps also be removed and everything < 1 minute shown as "just now". A minor issue with seconds is that it makes it super obvious that the value isn't updated while the dialog is visible, since it constantly says "5 seconds ago"

But this is quite a small thing, so any of these options (including the current one) are quite fine with me

Reply:
My initial confusion arises from the fact that I would imagine it would be almost impossible for students to click "submit" so quickly that it would be less than a full second

Reply:
Good points. Though it sounds a bit ridiculous (and not necessarily look good in the UI), I think showing seconds is useful for a user. I'd guess it's most often just seconds between a student being ready with editing the file and them making a submission.

Imagine you're a student who writes some code and is making a submission. It shows "just now" (meaning less than a minute ago). Then you remember that you didn't remove your "debug print" so you press cancel, remove that line and press "submit" button again. It took you about 10 seconds. It still shows "just now". Was the file saved? Is the latest change included in the submission? (Okay this was stupid but, still, my point holds. 😛)

Reply:
Well, the file is definitely saved since we save every file in the project on assignment submission :wink: Of course the student probably doesn't know that...

But I don't actually think the main benefit of this feature is "was the file saved?". I would say the main benefit of this feature is preventing submissions to the incorrect assignment. So when a student selects the wrong assignment and sees "last modified 2 hours ago" even though they just worked on it, then it becomes obvious that something is wrong. Selecting the incorrect assignment has happened in practice BTW, and if I recall correctly, feedback from a student who did so was specifically what prompted us to consider adding this feature in the first place.

Reply:
To be honest, I was targeting some other obscure case when doing this check for `<= 0 seconds` - a possibility that the file date is actually slightly in the future compared to current system time 😛
There are so many possible filesystems and each of them is handling last modification time in their own way so I wouldn't be surprised if this behaviour would be sometimes possible (file data/metadata caching or something) - this check would avoid a "modified -1 seconds ago" situation
Anyway, there seem to be following possibilities:
- drop "just now" altogether in favour of "0 seconds/minutes ago"
- reserve "just now" for a larger span of time (5 seconds? 30 seconds? a minute?)
- keep it as is

Reply:
The prosecution and the defense have presented their cases, you can be the judge on which is the best option of the three